### ISSUE TITLE
ReturnsSingle Return Option generates NullException

### ISSUE THREAD

**Issue by Jaxelr** [jonwagner/Insight.Database#281]
Hi,

Given the usage of the ReturnsSingle method as a Return option, the execution of the Mapping method MapChildren happens and the validation of the parent object fails since the ReturnsSingle makes sure a single instance is returned. This generates a NullException:

```c#
  IDbConnection conn = new SqlConnection("Server=localhost;database=master;Trusted_Connection=true;");

            conn.ExecuteSql(@"CREATE TABLE Parent (ParentId INT IDENTITY(1, 1) NOT NULL, name nvarchar(20) null);
                                CREATE TABLE Child (ChildId INT IDENTITY(1, 1) NOT NULL, ParentId int NOT NULL, Name nvarchar(50) null)");

            var response = conn.QuerySql(@" SET NOCOUNT ON;
                            --Parent
                            SELECT ParentId, Name FROM Parent WHERE @ParentId = ParentId
                            --Child
                            SELECT ParentId, ChildId, Name FROM Child WHERE @ParentId = ParentId", new { ParentId = 1 },
                Query.ReturnsSingle(Some<Parent>.Records)
                .ThenChildren(Some<Children>.Records, id: (d) => d.ParentId, into: (p, c) => p.Children = c));

```

But this works as expected:

```c#
            IDbConnection conn = new SqlConnection("Server=localhost;database=master;Trusted_Connection=true;");

            conn.ExecuteSql(@"CREATE TABLE Parent (ParentId INT IDENTITY(1, 1) NOT NULL, name nvarchar(20) null);
                                CREATE TABLE Child (ChildId INT IDENTITY(1, 1) NOT NULL, ParentId int NOT NULL, Name nvarchar(50) null)");

            var response = conn.QuerySql(@" SET NOCOUNT ON;
                            --Parent
                            SELECT ParentId, Name FROM Parent WHERE @ParentId = ParentId
                            --Child
                            SELECT ParentId, ChildId, Name FROM Child WHERE @ParentId = ParentId", new { ParentId = 1 },
                Query.ReturnsSingle(Some<Parent>.Records)
                .ThenChildren(Some<Children>.Records));
```

**Comment by Jaxelr** [comment]
UPDATE: This also doesnt work if using OneToOne instead of Some as the child mapper option.
@kevinsea rightiously pointed that i should include some fail tests. Here they are using SQL Server:

Test 1 Passes since the object is found.
Test 2 Fails since the mapping is executed incorrectly.
Test 3 Tries using OneToOne\<T\> instead of Some\<T\>.

```
    [TestFixture]
    public class InsightTest
    {

        IDbConnection conn = new SqlConnection("Server=localhost;database=master;Trusted_Connection=true;");

        public class Parent
        {
            public int ParentId { get; set; }
            public string Name { get; set; }
            public List<Children> Children { get; set; }
        }

        public class Children
        {
            public int ChildrenId { get; set; }
            public string Name { get; set; }
        }

        [Test]
        public void TestMethodFound()
        {
            try
            {
                conn.ExecuteSql(@"CREATE TABLE Parent (ParentId INT IDENTITY(1, 1) NOT NULL, name nvarchar(20) null);
                                CREATE TABLE Child (ChildId INT IDENTITY(1, 1) NOT NULL, ParentId int NOT NULL, Name nvarchar(50) null)
                                INSERT INTO Parent VALUES ('Parent')
                                INSERT INTO Child (ParentId, Name) SELECT ParentId, 'Child' FROM Child");

                var response = conn.QuerySql(@" SET NOCOUNT ON;
                            --Parent
                            SELECT ParentId, Name FROM Parent WHERE @ParentId = ParentId
                            --Child
                            SELECT ParentId, ChildId, Name FROM Child WHERE @ParentId = ParentId", new { ParentId = 1 },
                    Query.ReturnsSingle(Some<Parent>.Records)
                    .ThenChildren(OneToOne<Children>.Records, id: (d) => d.ParentId, into: (p, c) => p.Children = c));

                Assert.AreEqual(response.ParentId, 1);
                Assert.AreEqual(response.Name, "Parent");

            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }
            finally
            {
                conn.ExecuteSql(@"DROP TABLE Parent; Drop Table Child");
            }
        }

        [Test]
        public void TestMethodNotFound()
        {
            try
            {
                conn.ExecuteSql(@"CREATE TABLE Parent (ParentId INT IDENTITY(1, 1) NOT NULL, name nvarchar(20) null);
                                CREATE TABLE Child (ChildId INT IDENTITY(1, 1) NOT NULL, ParentId int NOT NULL, Name nvarchar(50) null)
                                INSERT INTO Parent VALUES ('Parent')
                                INSERT INTO Child (ParentId, Name) SELECT ParentId, 'Child' FROM Child");

                var response = conn.QuerySql(@" SET NOCOUNT ON;
                            --Parent
                            SELECT ParentId, Name FROM Parent WHERE @ParentId = ParentId
                            --Child
                            SELECT ParentId, ChildId, Name FROM Child WHERE @ParentId = ParentId", new { ParentId = 2 },
                    Query.ReturnsSingle(Some<Parent>.Records)
                    .ThenChildren(Some<Children>.Records, id: (d) => d.ParentId, into: (p, c) => p.Children = c));

                Assert.AreEqual(response, null);

            }
            catch (Exception e)
            {
                Assert.Fail(e.Message);
            }
            finally
            {
                conn.ExecuteSql(@"DROP TABLE Parent; Drop Table Child");
            }
        }

        [Test]
        public void TestMethodNotFoundOneToOne()
        {
            try
            {
                conn.ExecuteSql(@"CREATE TABLE Parent (ParentId INT IDENTITY(1, 1) NOT NULL, name nvarchar(20) null);
                                CREATE TABLE Child (ChildId INT IDENTITY(1, 1) NOT NULL, ParentId int NOT NULL, Name nvarchar(50) null)
                                INSERT INTO Parent VALUES ('Parent')
                                INSERT INTO Child (ParentId, Name) SELECT ParentId, 'Child' FROM Child");

                var response = conn.QuerySql(@" SET NOCOUNT ON;
                            --Parent
                            SELECT ParentId, Name FROM Parent WHERE @ParentId = ParentId
                            --Child
                            SELECT ParentId, ChildId, Name FROM Child WHERE @ParentId = ParentId", new { ParentId = 2 },
                    Query.ReturnsSingle(Some<Parent>.Records)
                    .ThenChildren(OneToOne<Children>.Records, id: (d) => d.ParentId, into: (p, c) => p.Children = c));

                Assert.AreEqual(response, null);

            }
            catch (Exception e)
            {
                Assert.Fail(e.Message);
            }
            finally
            {
                conn.ExecuteSql(@"DROP TABLE Parent; Drop Table Child");
            }
        }
    }
```

**Comment by jonwagner** [comment]
Merged into build 5.2.9 (finally!)
